src/tasks/LibraryBuilder/LibraryBuilder.cs

```diff
@@ -226,6 +253,79 @@ private static void WriteLinkerScriptFile(string exportsFile, List<string> expor
                 .Replace("%GLOBAL_SYMBOLS%", globalExports));
     }
 
+    private void WriteAutoInitializationFromTemplate()
+    {
+        string appContextEnvVariables = GenerateAppContextEnvVariables();
+        string assembliesLoader = GenerateAssembliesLoader();
+        string assetsPath = GenerateAssetsPath();
+        string runtimeConfig = GenerateRuntimeConfig();
+        File.WriteAllText(Path.Combine(OutputDirectory, "autoinit.c"),
+            Utils.GetEmbeddedResource("autoinit.c")
+                .Replace("%APPCTX_ENV_VARIABLES%", appContextEnvVariables)
+                .Replace("%ASSEMBLIES_LOADER%", assembliesLoader)
+                .Replace("%ASSETS_PATH%", assetsPath)
+                .Replace("%RUNTIME_CONFIG%", runtimeConfig));
+    }
+
+    private string GenerateAssembliesLoader()
+    {
+        var assembliesLoader = new StringBuilder();
+        foreach (string exportedAssembly in exportedAssemblies)
+        {
+            assembliesLoader.AppendLine($"    mono_assembly_open(\"{exportedAssembly}\", NULL);");
```

Comment:
Since loading the assembly first makes sure everything initializes properly, having a function would give the same ability to custom initializers.  On the other hand, this makes things much more smooth.  Come to think of it, why don't we have a function that just does what you're doing in assembly_list.c?

Reply:
We could move the `mono_assembly_open` logic in `assembly_list.c` (and probably rename it to like `load_assemblies`), that would just be refactoring the code if it helps to have that separation. On the other hand, if there are no assemblies to load (no UnmanagedCallersOnly decorated functions) then that function would be an empty function yet still be called at the end of `runtime_init_callback`.

What do we anticipate as the number of assemblies that the end user will need to be loaded at the end of the runtime initialization? Is it in the order of 1-10 or 10-50? If its on the larger side, I think having the `mono_assembly_open` in a separate file could be a clear separation, but does the separation need to be in 2 files?

Reply:
Moved `mono_assembly_open` into the separate function (now called `load_assemblies_with_exported_symbols`) in the separate source file (now called `load_assemblies.c`) and replaced the previous logic that generated a function that returned a list of assemblies.

The change is specifically in the commit 1Move assembly loading into separate function`
https://github.com/mdh1418/runtime/commit/3b71e955501c32489a01dfe503e40898da367152